File: frontend/src/lib/supplier.ts
// 供应商主数据前端类型 + API。对齐后端 schemas/supplier.py。
import { api } from "./api";
import type { Page } from "./catalog";

export type SupplierStatus = "ACTIVE" | "INACTIVE";

/** 状态徽标映射(镜像 UI 呈现,非业务规则)。ACTIVE=启用 / INACTIVE=停用。 */
export const SUPPLIER_STATUS_META: Record<SupplierStatus, { label: string; color: string }> = {
  ACTIVE: { label: "启用", color: "success" },
  INACTIVE: { label: "停用", color: "default" },
};

export interface SupplierOut {
  id: number;
  code: string;
  name: string;
  default_currency: string | null;
  contact_name: string | null;
  contact_phone: string | null;
  contact_email: string | null;
  address: string | null;
  status: SupplierStatus;
}

export interface SupplierListItem {
  id: number;
  code: string;
  name: string;
  default_currency: string | null;
  contact_name: string | null;
  status: SupplierStatus;
  updated_at: string;
}

/** 建 / 改供应商入参(POST 与 PUT 同体)。 */
export interface SupplierSaveBody {
  name: string;
  default_currency?: string | null;
  contact_name?: string | null;
  contact_phone?: string | null;
  contact_email?: string | null;
  address?: string | null;
}

export interface SupplierListFilters {
  status?: string;
  q?: string;
  page?: number;
  size?: number;
}

function qs(p: Record<string, unknown>): string {
  const q = new URLSearchParams();
  Object.entries(p).forEach(([k, v]) => {
    if (v !== undefined && v !== null && v !== "") q.set(k, String(v));
  });
  const s = q.toString();
  return s ? `?${s}` : "";
}

export const supplierApi = {
  list: (p: SupplierListFilters) =>
    api.get<Page<SupplierListItem>>(`/api/v1/suppliers${qs(p as Record<string, unknown>)}`),
  get: (id: number) => api.get<SupplierOut>(`/api/v1/suppliers/${id}`),
  create: (b: SupplierSaveBody) => api.post<SupplierOut>("/api/v1/suppliers", b),
  update: (id: number, b: SupplierSaveBody) => api.put<SupplierOut>(`/api/v1/suppliers/${id}`, b),
  activate: (id: number) => api.post<SupplierOut>(`/api/v1/suppliers/${id}/activate`),
  deactivate: (id: number) => api.post<SupplierOut>(`/api/v1/suppliers/${id}/deactivate`),
};

File: backend/app/schemas/supplier.py
"""供应商 schemas(照 customers 档次:最小主数据 + 默认采购币种 + 启停)。"""
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field, field_validator


def _valid_currency(v: str | None) -> str | None:
    if v is not None and not (len(v) == 3 and v.isalpha() and v.isupper()):
        raise ValueError("default_currency 必须是 ISO4217 三字母大写 code(如 USD)")
    return v


class SupplierCreateIn(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    default_currency: str | None = None
    contact_name: str | None = Field(default=None, max_length=100)
    contact_phone: str | None = Field(default=None, max_length=30)
    contact_email: str | None = Field(default=None, max_length=255)
    address: str | None = None

    _v_currency = field_validator("default_currency")(_valid_currency)


class SupplierUpdateIn(SupplierCreateIn):
    """编辑:同建单字段(不含 code/status,后者走 activate/deactivate)。"""


class SupplierOut(BaseModel):
    id: int
    code: str
    name: str
    default_currency: str | None
    contact_name: str | None
    contact_phone: str | None
    contact_email: str | None
    address: str | None
    status: str


class SupplierListItem(BaseModel):
    id: int
    code: str
    name: str
    default_currency: str | None
    contact_name: str | None
    status: str
    updated_at: datetime
